File: scripts/clean.ts
#!/usr/bin/env -S node --no-warnings --experimental-strip-types
import { rm } from "node:fs/promises";
import { $ } from "zx";
import { abs } from "./_utils.ts";

$.cwd = abs();
$.verbose = true;
$.env["SKIPPER_KUBE_CONTEXT"] ??= "orbstack";

await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete namespace skipper-development --ignore-not-found`;
await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete namespace skipper-development-fixtures --ignore-not-found`;
await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete namespace skipper-test --ignore-not-found`;
await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete namespace skipper-test-fixtures --ignore-not-found`;
await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete namespace otel-lgtm --ignore-not-found`;
await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete -f deploy/cluster/metrics-server.yaml --ignore-not-found`;
await $`kubectl --context="$SKIPPER_KUBE_CONTEXT" delete -f deploy/kube-system/metrics-server.yaml --ignore-not-found`;

await rm(abs("tmp/krane"), { recursive: true, force: true });
await rm(abs("tmp/logs"), { recursive: true, force: true });
await rm(abs("tmp/paseto"), { recursive: true, force: true });
await rm(abs("tmp/test"), { recursive: true, force: true });

File: cmd/dev/clean.go
package main

import (
	"errors"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/gadget-inc/skipper/internal/cmd"
	"github.com/gadget-inc/skipper/internal/dev/exec"
	"github.com/spf13/cobra"
)

// newCleanCmd deletes every Kubernetes namespace skipper deploys plus
// the local krane / paseto / log / test scratch directories.
func newCleanCmd() *cobra.Command {
	return cmd.Build(cmd.Spec{
		Use:   "clean",
		Short: "Tear down skipper's K8s namespaces and local scratch dirs",
		RunE: func(c *cobra.Command, args []string) error {
			ctx := c.Context()
			kubectx := os.Getenv("SKIPPER_KUBECTL_CONTEXT")
			if kubectx == "" {
				kubectx = "orbstack"
			}

			namespaces := []string{
				"skipper-development",
				"skipper-development-fixtures",
				"skipper-test",
				"skipper-test-fixtures",
				"otel-lgtm",
			}
			for _, ns := range namespaces {
				_ = exec.Run(ctx, "kubectl",
					[]string{"--context=" + kubectx, "delete", "namespace", ns, "--ignore-not-found"},
				)
			}

			manifests := []string{
				"deploy/cluster/metrics-server.yaml",
				"deploy/kube-system/metrics-server.yaml",
			}
			for _, f := range manifests {
				_ = exec.Run(ctx, "kubectl",
					[]string{"--context=" + kubectx, "delete", "-f", f, "--ignore-not-found"},
				)
			}

			workspace := os.Getenv("WORKSPACE_DIR")
			scratch := []string{"tmp/krane", "tmp/logs", "tmp/paseto", "tmp/test"}
			for _, sub := range scratch {
				path := filepath.Join(workspace, sub)
				if err := os.RemoveAll(path); err != nil && !errors.Is(err, fs.ErrNotExist) {
					return err
				}
			}
			return nil
		},
	})
}
